pkg/services/ngalert/migration/service.go

```diff
@@ -486,9 +490,18 @@ func (ms *migrationService) migrateAllOrgs(ctx context.Context) error {
 			return err
 		}
 
-		if len(om.silences) > 0 {
-			om.log.Debug("Writing silences file", "silences", len(om.silences))
-			if err := writeSilencesFile(ms.cfg.DataPath, o.ID, om.silences); err != nil {
+		var silences []*pb.MeshSilence
+		if om.rulesWithErrorSilenceLabels > 0 {
+			om.log.Info("Creating silence for rules with ExecutionErrorState = keep_state", "rules", om.rulesWithErrorSilenceLabels)
+			silences = append(silences, errorSilence())
+		}
+		if om.rulesWithNoDataSilenceLabels > 0 {
+			om.log.Info("Creating silence for rules with NoDataState = keep_state", "rules", om.rulesWithNoDataSilenceLabels)
+			silences = append(silences, noDataSilence())
+		}
+		if len(silences) > 0 {
+			om.log.Debug("Writing silences file", "silences", len(silences))
+			if err := ms.writeSilencesFile(om.orgID, silences); err != nil {
 				return fmt.Errorf("write silence file for org %d: %w", o.ID, err)
 			}
 		}
```

Comment:
There's enough here that it probably shouldn't be under `migrateAllOrgs` anymore, especially since the logic is actually being tested in `silences_test.go`. WDYT about moving this block to its own function/method in `silences.go` and testing that directly?